sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -353,7 +353,7 @@ case class DataSource(
       case (dataSource: RelationProvider, Some(schema)) =>
         val baseRelation =
           dataSource.createRelation(sparkSession.sqlContext, caseInsensitiveOptions)
-        if (baseRelation.schema != schema) {
+        if (baseRelation.schema.withoutMetadata != schema.withoutMetadata) {
```

Comment:
seems we can just call `DataType.equalsIgnoreCompatibleNullability(baseRelation.schema, schema)`

Reply:
It ignore `nullable`, It is possible have a chance to occur failure in query later?  For example, In old 3.1, nullable is true, and there are some null value actually, But nullable is false now, I'm not sure that if have a chance to occur failure of the query.

Reply:
I also think we shouldn't ignore `nullable`.

Reply:
The user-specified schema is ignored at the end. Technically we can completely ignore it and don't do any checks. The check is to make sure the user-specified schema is not too wrong. If we can ignore the metadata, it seems to me that compatible nullability can also be ignored. e.g. the actual column is not nullable but is marked as nullable in the user-specified schema.

Reply:
Agree. Thanks for the explanation.